Index: tools/clang/lib/Basic/FileManager.cpp
===================================================================
--- tools/clang/lib/Basic/FileManager.cpp	(revision 160036)
+++ tools/clang/lib/Basic/FileManager.cpp	(working copy)
@@ -112,7 +112,13 @@
 
   size_t size() const { return UniqueFiles.size(); }
 
-  void erase(const FileEntry *Entry) { UniqueFiles.erase(Entry->getName()); }
+  void erase(const FileEntry *Entry) {
+    std::string FullPath(GetFullPath(Entry->getName()));
+
+    // Lowercase string because Windows filesystem is case insensitive.
+    FullPath = StringRef(FullPath).lower();
+    UniqueFiles.erase(FullPath);
+  }
 };
 
 //===----------------------------------------------------------------------===//
