Index: tools/clang/lib/Frontend/VerifyDiagnosticConsumer.cpp
===================================================================
--- tools/clang/lib/Frontend/VerifyDiagnosticConsumer.cpp	(revision 159991)
+++ tools/clang/lib/Frontend/VerifyDiagnosticConsumer.cpp	(working copy)
@@ -527,6 +527,9 @@
 
   // Reset the buffer, we have processed all the diagnostics in it.
   Buffer.reset(new TextDiagnosticBuffer());
+  ED.Errors.clear();
+  ED.Warnings.clear();
+  ED.Notes.clear();
 }
 
 DiagnosticConsumer *
