Index: tools/Makefile
===================================================================
--- tools/Makefile	(revision 135054)
+++ tools/Makefile	(working copy)
@@ -14,7 +14,7 @@
 
 # Build LLDB if present. Note LLDB must be built last as it depends on the
 # wider LLVM infrastructure (including Clang).
-OPTIONAL_DIRS := lldb
+OPTIONAL_DIRS := lldb cling
 
 # NOTE: The tools are organized into five groups of four consisting of one
 # large and three small executables. This is done to minimize memory load
