Index: test/Makefile
===================================================================
--- test/Makefile	(revision 153954)
+++ test/Makefile	(working copy)
@@ -61,6 +61,8 @@
 ifndef TESTSUITE
 ifeq ($(shell test -f $(PROJ_OBJ_DIR)/../tools/clang/Makefile && echo OK), OK)
 LIT_ALL_TESTSUITES += $(PROJ_OBJ_DIR)/../tools/clang/test
+ifeq ($(shell test -f $(PROJ_OBJ_DIR)/../tools/cling/Makefile && echo OK), OK)
+LIT_ALL_TESTSUITES += $(PROJ_OBJ_DIR)/../tools/cling/test
 
 # Force creation of Clang's lit.site.cfg.
 clang-lit-site-cfg: FORCE
