Index: test/Makefile
===================================================================
--- test/Makefile	(revision 154169)
+++ test/Makefile	(working copy)
@@ -67,8 +67,17 @@
 	$(MAKE) -C $(PROJ_OBJ_DIR)/../tools/clang/test lit.site.cfg Unit/lit.site.cfg
 extra-lit-site-cfgs:: clang-lit-site-cfg
 endif
+
+ifeq ($(NOCLING)$(shell test -f $(PROJ_OBJ_DIR)/../tools/cling/Makefile && echo OK), OK)
+LIT_ALL_TESTSUITES += $(PROJ_OBJ_DIR)/../tools/cling/test
+
+# Force creation of Cling's lit.site.cfg.
+cling-lit-site-cfg: FORCE
+	$(MAKE) -C $(PROJ_OBJ_DIR)/../tools/cling/test lit.site.cfg
+extra-lit-site-cfgs:: cling-lit-site-cfg
 endif
 endif
+endif
 
 IGNORE_TESTS :=
 
