--- CMakeLists.txt.orig	2014-07-03 15:21:44.000000000 +0200
+++ CMakeLists.txt	2014-07-03 15:22:40.000000000 +0200
@@ -80,10 +80,10 @@
 add_definitions( -D_FILE_OFFSET_BITS=64)
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CXX11_FLAG_ENABLE}")
 
+include_directories( ${Boost_INCLUDE_DIRS})
 include_directories( ${HTTPLIB_PKG_INCLUDE_DIRS} ${GLIB2_PKG_INCLUDE_DIRS} ${CMAKE_CURRENT_SOURCE_DIR}/deps ${CMAKE_CURRENT_SOURCE_DIR}/deps/libneon/src/ )
 include_directories( ${CMAKE_SOURCE_DIR}/include/davix )
 include_directories( ${CMAKE_SOURCE_DIR}/src/libs/)
-include_directories( ${Boost_INCLUDE_DIRS})
 
 add_subdirectory (src)
 add_subdirectory (doc)
