diff --git a/CMakeLists.txt b/CMakeLists.txt
index 56f830c..b1d308c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -37,6 +37,9 @@ set(VERSION_TAG "std")
 # dependency mode
 option(BOOST_EXTERNAL "Enable the usage of an external boost dependency" TRUE)

+# components
+option(ENABLE_TOOLS "enable or disable tools " TRUE)
+
 # features
 option(ENABLE_IPV6 "enable or disable IPv6 support " TRUE)
 option(ENABLE_TCP_NODELAY "enable or disable tcp_nodelay" TRUE)
@@ -78,10 +81,10 @@ add_definitions( -D__DAVIX_INSIDE__)         # protection flag
 add_definitions( -D__STDC_LIMIT_MACROS )     # C99 limit macros
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CXX11_FLAG_ENABLE}")

+include_directories( ${Boost_INCLUDE_DIRS})
 include_directories( ${HTTPLIB_PKG_INCLUDE_DIRS} ${GLIB2_PKG_INCLUDE_DIRS} ${CMAKE_CURRENT_SOURCE_DIR}/deps ${CMAKE_CURRENT_SOURCE_DIR}/deps/libneon/src/ )
 include_directories( ${CMAKE_SOURCE_DIR}/include/davix )
 include_directories( ${CMAKE_SOURCE_DIR}/src/libs/)
-include_directories( ${Boost_INCLUDE_DIRS})

 add_subdirectory (src)
 add_subdirectory (doc)
diff --git a/deps/libneon/src/CMakeLists.txt b/deps/libneon/src/CMakeLists.txt
index a412fee..1eb67db 100644
--- a/deps/libneon/src/CMakeLists.txt
+++ b/deps/libneon/src/CMakeLists.txt
@@ -16,4 +16,9 @@ endif(HAVE_GNUTLS)

 add_library(neon STATIC ${src_neon} ${src_ssl})

+# if the static library is produced, then we need also to install the dependent libraries to be useful
+if(STATIC_LIBRARY)
+  install(TARGETS neon DESTINATION ${LIB_INSTALL_DIR} )
+endif()
+

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index e318b90..ed341d0 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -28,7 +28,9 @@ link_directories(${HTTPLIB_PKG_LIBRARY_DIR} ${CRYPTO_PKG_LIBRARY_DIR}
 add_subdirectory(copy)
 add_subdirectory(libs)
 add_subdirectory(xml)
-add_subdirectory(tools)
+if(ENABLE_TOOLS)
+  add_subdirectory(tools)
+endif()

 add_subdirectory(neon)
 SET(SRC_HTTP ${src_neon})
@@ -64,6 +66,9 @@ add_PkgConfigFile_for_Library("davix.pc" libdavix
                                CFLAGS " -D_FILE_OFFSET_BITS=64 "
                                DESCRIPTION "Libdavix, IO library for HTTP based protocols"
                                )
+#install pkgconfig
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/davix.pc
+        DESTINATION ${PKGCONFIG_FILES_DIR} )

 endif(SHARED_LIBRARY)

@@ -86,8 +91,4 @@ endif(STATIC_LIBRARY)
 SET(davix_tool_common_src "${davix_tool_common_src_up}" PARENT_SCOPE)


-#install pkgconfig
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/davix.pc
-        DESTINATION ${PKGCONFIG_FILES_DIR} )
-

diff --git a/deps/boost_intern/src/CMakeLists.txt b/src/CMakeLists.txt
index 6e169bb..2426393 100644
--- a/deps/boost_intern/src/CMakeLists.txt
+++ b/deps/boost_intern/src/CMakeLists.txt
@@ -18,4 +18,9 @@ add_definitions( -fPIC )
 add_definitions(-fvisibility=hidden)
 add_library(boost_static_internal STATIC ${SRC_BOOST_ALL})

+# if the static library is produced, then we need also to install the dependent libraries to be useful
+if(STATIC_LIBRARY)
+  install(TARGETS boost_static_internal DESTINATION ${LIB_INSTALL_DIR} )
+endif()
+
 endif(NOT Boost_FOUND)
