From a548eb755121c1d4afb8b788c27b653ad6567d5a Mon Sep 17 00:00:00 2001
From: ferdymercury <ferdymercury@users.noreply.github.com>
Date: Wed, 22 Apr 2026 19:46:28 +0200
Subject: [PATCH] message libs

---
 src/CMakeLists.txt | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 43f3a77e..eab761fb 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -56,6 +56,8 @@ if (CIVETWEB_ENABLE_SSL)
     find_package(OpenSSL)
     include_directories(${OPENSSL_INCLUDE_DIR})
     message(STATUS "OpenSSL include directory: ${OPENSSL_INCLUDE_DIR}")
+    file(GLOB_RECURSE MyHeaders ${OPENSSL_INCLUDE_DIR}/*.h)
+    message(WARNING ${MyHeaders})
     target_link_libraries(civetweb-c-library ${OPENSSL_LIBRARIES})
   endif()
 endif()
-- 
2.34.1

