From ecb6e72a7be25f3f8bfc1a0b507ccae91fa5bf4f Mon Sep 17 00:00:00 2001
From: ferdymercury <ferdymercury@users.noreply.github.com>
Date: Wed, 22 Apr 2026 16:42:49 +0200
Subject: [PATCH] fix inc path

---
 src/CMakeLists.txt | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 43f3a77e..7d25bfd5 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -53,6 +53,7 @@ if (CIVETWEB_ENABLE_SSL)
       target_link_libraries(civetweb-c-library -ldl)
     endif()
   else()
+    include_directories(${OPENSSL_INCLUDE_DIR})
     find_package(OpenSSL)
     include_directories(${OPENSSL_INCLUDE_DIR})
     message(STATUS "OpenSSL include directory: ${OPENSSL_INCLUDE_DIR}")
-- 
2.34.1

