--- /dev/null	2026-03-12 10:46:44
+++ CMakeLists.txt	2026-03-12 10:46:24
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.20)
+project(lz4 VERSION 1.0) # The version does not matter here
+
+add_library(lz4-static STATIC
+    lib/lz4.c
+    lib/lz4frame.c
+    lib/lz4hc.c
+)
+
+target_sources(lz4-static PUBLIC
+    FILE_SET HEADERS
+    BASE_DIRS ./lib
+    FILES lib/lz4.h lib/lz4frame.h lib/lz4hc.h
+)
+
+target_include_directories(lz4-static PUBLIC ./)
+
+# Install the target and its file sets
+install(TARGETS lz4-static
+    FILE_SET HEADERS DESTINATION ${CMAKE_INSTALL_PREFIX}/include
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_PREFIX}/lib
+)
+
