diff --git a/bindings/pyroot_experimental/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx b/bindings/pyroot_experimental/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx
index 3559a45..a4f7a88 100644
--- a/bindings/pyroot_experimental/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx
+++ b/bindings/pyroot_experimental/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx
@@ -109,6 +109,10 @@ namespace {
 class ApplicationStarter {
 public:
     ApplicationStarter() {
+    // Insure ROOT's atexit is executed *after* the atexit that calls
+    // ApplicationStarter's destructor, by forcing the ROOT's atexit
+    // registration now.
+        TROOT::Initialize();
     // setup dummy holders for global and std namespaces
         assert(g_classrefs.size() == GLOBAL_HANDLE);
         g_name2classrefidx[""]      = GLOBAL_HANDLE;
