diff --git a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
index aec153242b..abcfd33597 100644
--- a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
+++ b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
@@ -336,6 +336,13 @@ namespace {
         }
   #endif // _LIBCPP_VERSION
 
+  // special case for Mac, use `xcrun` to get around various recent path changes
+  // in system updates
+  #ifdef __APPLE__
+        if (sArguments.empty())
+          ReadCompilerIncludePaths("xcrun c++", buffer, sArguments, Verbose);
+  #endif
+
   // First try the relative path 'g++'
   #ifdef CLING_CXX_RLTV
         if (sArguments.empty())
