diff --git a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
index f4a61b6640..9956a11bdf 100644
--- a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
+++ b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
@@ -114,11 +114,11 @@ namespace {
                                        bool Verbose) {
     std::string CppInclQuery("LC_ALL=C ");
     CppInclQuery.append(Compiler);
 
     CppInclQuery.append(" -xc++ -E -v /dev/null 2>&1 |"
-                        " sed -n -e '/^.include/,${' -e '/^ \\/.*++/p' -e '}'");
+                        " sed -n -e '/^.*include/,${' -e '/^ \\/.*++/p' -e '}'");
 
     if (Verbose)
       cling::log() << "Looking for C++ headers with:\n  " << CppInclQuery << "\n";
 
     if (FILE *PF = ::popen(CppInclQuery.c_str(), "r")) {
