diff --git a/src/builtins/zlib/adler32_cf.c b/src/builtins/zlib/adler32_cf.c
index 7cbf8a030a..bc21d2d066 100644
--- a/src/builtins/zlib/adler32_cf.c
+++ b/src/builtins/zlib/adler32_cf.c
@@ -171,7 +171,7 @@ uLong ZEXPORT adler32_default(uLong adler, const Bytef *buf, uInt len)
     return adler | (sum2 << 16);
 }
 
-#if defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
+#if defined (NOT_DEFINED) && defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
 
 #ifdef _MSC_VER
 
diff --git a/src/builtins/zlib/deflate_cf.c b/src/builtins/zlib/deflate_cf.c
index 54aebec62a..6ac7ecdfd0 100644
--- a/src/builtins/zlib/deflate_cf.c
+++ b/src/builtins/zlib/deflate_cf.c
@@ -1165,7 +1165,7 @@ static uint32_t hash_func(deflate_state *s, uint32_t h, void* str) {
 }
 
 #endif // ARMv8 without crc32 support
-#elif defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
+#elif defined (NOT_DEFINED) && defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
 
 static uint32_t hash_func_sse42(deflate_state *s, uint32_t UNUSED(h), void* str) __attribute__ ((__target__ ("sse4.2")));
 
@@ -2477,7 +2477,7 @@ static void fill_window_default(s)
            "not enough room for search");
 }
 
-#if defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
+#if defined (NOT_DEFINED) && defined (__x86_64__) && defined (__linux__) && ((__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9)) || (__clang__))
 
 /* ===========================================================================
  * Fill the window when the lookahead becomes insufficient.
