--- /home/axel/build/llvm-orig/src/./tools/clang/lib/AST/VTableBuilder.cpp	2013-04-11 09:34:54.273161777 +0200
+++ ./tools/clang/lib/AST/VTableBuilder.cpp	2013-04-24 19:40:14.521496495 +0200
@@ -1165,7 +1165,8 @@
       continue;
     }
 
-    if (MD->getParent() == MostDerivedClass)
+    if (MD->getParent()->getCanonicalDecl()
+        == MostDerivedClass->getCanonicalDecl())
       AddThunk(MD, Thunk);
   }
 }
@@ -1362,7 +1363,8 @@
   
   // If the overrider is the first base in the primary base chain, we know
   // that the overrider will be used.
-  if (Overrider->getParent() == FirstBaseInPrimaryBaseChain)
+  if (Overrider->getParent()->getCanonicalDecl()
+      == FirstBaseInPrimaryBaseChain->getCanonicalDecl())
     return true;
   
   VTableBuilder::PrimaryBasesSetVectorTy PrimaryBases;
@@ -1426,7 +1428,8 @@
       const CXXMethodDecl *OverriddenMD = *I;
       
       // We found our overridden method.
-      if (OverriddenMD->getParent() == PrimaryBase)
+      if (OverriddenMD->getParent()->getCanonicalDecl()
+          == PrimaryBase->getCanonicalDecl())
         return OverriddenMD;
     }
   }
@@ -1521,7 +1524,8 @@
                                   Overrider);
 
           if (ThisAdjustment.VCallOffsetOffset &&
-              Overrider.Method->getParent() == MostDerivedClass) {
+              Overrider.Method->getParent()->getCanonicalDecl()
+              == MostDerivedClass->getCanonicalDecl()) {
 
             // There's no return adjustment from OverriddenMD and MD,
             // but that doesn't mean there isn't one between MD and
