--- ./tools/clang/include/clang/AST/Type.h	2013-04-11 09:35:12.320974197 +0200
+++ ./tools/clang/include/clang/AST/Type.h	2013-04-24 19:40:12.589511759 +0200
@@ -2959,7 +2959,7 @@
   }
   bool isNothrow(ASTContext &Ctx) const {
     ExceptionSpecificationType EST = getExceptionSpecType();
-    assert(EST != EST_Unevaluated && EST != EST_Uninstantiated);
+    //AXEL: assert(EST != EST_Unevaluated && EST != EST_Uninstantiated);
     if (EST == EST_DynamicNone || EST == EST_BasicNoexcept)
       return true;
     if (EST != EST_ComputedNoexcept)
