--- ./tools/clang/tools/Makefile	2013-04-11 09:35:14.440952162 +0200
+++ ./tools/clang/tools/Makefile	2013-04-24 19:40:13.881501551 +0200
@@ -11,8 +11,12 @@
 
 include $(CLANG_LEVEL)/../../Makefile.config
 
+ifneq ($(ROOT_BUILDCLANG),)
 DIRS := driver libclang c-index-test arcmt-test c-arcmt-test diagtool \
         clang-check clang-format
+else
+DIRS := libclang
+endif
 
 # Recurse into the extra repository of tools if present.
 OPTIONAL_DIRS := extra
--- ./tools/clang/CMakeLists.txt	2013-04-11 09:35:15.232943930 +0200
+++ ./tools/clang/CMakeLists.txt	2013-04-24 19:40:15.789486476 +0200
@@ -286,7 +286,6 @@
        ${LLVM_INCLUDE_TESTS})
 
 # TODO: docs.
-add_subdirectory(test)
 
 if( CLANG_INCLUDE_TESTS )
   add_subdirectory(unittests)
--- ./tools/clang/Makefile	2012-10-18 20:43:31.328406919 +0200
+++ ./tools/clang/Makefile	2013-04-24 19:40:15.789486476 +0200
@@ -95,9 +95,6 @@
 report::
 	@ $(MAKE) -C test report
 
-clean::
-	@ $(MAKE) -C test clean
-
 libs-only: all
 
 tags::
--- ./tools/clang/unittests/Makefile	2013-01-09 12:55:41.531207777 +0100
+++ ./tools/clang/unittests/Makefile	2013-04-24 19:40:13.921501235 +0200
@@ -1,42 +1,3 @@
-##===- unittests/Makefile ----------------------------------*- Makefile -*-===##
-#
-#                     The LLVM Compiler Infrastructure
-#
-# This file is distributed under the University of Illinois Open Source
-# License. See LICENSE.TXT for details.
-#
-##===----------------------------------------------------------------------===##
-
-# If CLANG_LEVEL is not set, then we are the top-level Makefile. Otherwise, we
-# are being included from a subdirectory makefile.
-
-ifndef CLANG_LEVEL
-
-IS_UNITTEST_LEVEL := 1
-CLANG_LEVEL := ..
-PARALLEL_DIRS = Basic Lex
-
-include $(CLANG_LEVEL)/../..//Makefile.config
-
-ifeq ($(ENABLE_CLANG_REWRITER),1)
-PARALLEL_DIRS += Format
-endif
-
-ifeq ($(ENABLE_CLANG_REWRITER),1)
-PARALLEL_DIRS += ASTMatchers AST Tooling
-endif
-
-ifeq ($(ENABLE_CLANG_STATIC_ANALYZER),1)
-PARALLEL_DIRS += Frontend
-endif
-
-endif  # CLANG_LEVEL
-
-include $(CLANG_LEVEL)/Makefile
-
-ifndef IS_UNITTEST_LEVEL
-
-MAKEFILE_UNITTEST_NO_INCLUDE_COMMON := 1
-include $(LLVM_SRC_ROOT)/unittests/Makefile.unittest
-
-endif  # IS_UNITTEST_LEVEL
+all:
+clean:
+install:
--- ./unittests/Makefile	2013-01-09 12:53:18.572818697 +0100
+++ ./unittests/Makefile	2013-04-24 19:40:16.621479902 +0200
@@ -7,11 +7,10 @@
 #
 ##===----------------------------------------------------------------------===##
 
-LEVEL = ..
+all:;
 
-PARALLEL_DIRS = ADT ExecutionEngine Support Transforms IR Analysis Bitcode
 
-include $(LEVEL)/Makefile.common
+clean:;
 
-clean::
-	$(Verb) $(RM) -f *Tests
+
+install:;
