--- ./tools/clang/lib/Basic/SourceManager.cpp	2013-04-11 09:34:53.729167430 +0200
+++ ./tools/clang/lib/Basic/SourceManager.cpp	2013-04-24 19:40:14.089499908 +0200
@@ -1962,8 +1962,8 @@
   // built-in is before non-built-in
   if (LIsBuiltins != RIsBuiltins)
     return LIsBuiltins;
-  assert(LIsBuiltins && RIsBuiltins &&
-         "Non-built-in locations must be rooted in the main file");
+  //assert(LIsBuiltins && RIsBuiltins &&
+  //       "Non-built-in locations must be rooted in the main file");
   // Both are in built-in buffers, but from different files. We just claim that
   // lower IDs come first.
   return LOffs.first < ROffs.first;
